go/types/sequence_cursor.go

```diff
@@ -4,24 +4,41 @@
 
 package types
 
-import "github.com/attic-labs/noms/go/d"
+import (
+	"math"
+	"os"
+	"runtime"
+
+	"github.com/attic-labs/noms/go/d"
+)
 
 // sequenceCursor explores a tree of sequence items.
 type sequenceCursor struct {
-	parent *sequenceCursor
-	seq    sequence
-	idx    int
+	parent        *sequenceCursor
+	seq           sequence
+	idx           int
+	readAheadMap  map[int]chan sequence
+	readAheadNext int
 }
 
+// newSequenceCursor creates a cursor on seq positioned at idx.
+// If idx < 0, count backward from the end of seq.
 func newSequenceCursor(parent *sequenceCursor, seq sequence, idx int) *sequenceCursor {
 	d.PanicIfTrue(seq == nil)
 	if idx < 0 {
 		idx += seq.seqLen()
 		d.PanicIfFalse(idx >= 0)
 	}
 
-	cur := &sequenceCursor{parent, seq, idx}
-	return cur
+	var readAheadMap map[int]chan sequence
+	// enable read-ahead if the cursor represents to a meta-sequence containing leaves
+	ms, ok := seq.(metaSequence)
+	if os.Getenv("NOMS_NO_READAHEAD") == "" && ok {
+		if ms.tuples[0].ref.height == 1 {
```

Comment:
Ref height isn't enough unfortunately, since it's the height of the ref, not the prolly-tree.

This can be different from the tree height if the leaves are refs have heights > 1. You could get this with any nested collections, or just a collection of refs.

One way you could fix this would be to have a `seqHeight` property on `sequenceCursor` which you set from the 2 places sequenceCursors are constructed from: `newCursorAt` and `newCursorAtIndex` (and `clone()` I suppose). By stepping up the cursor chain after it's constructed.